### PR TITLE
qemu_guest_agent: update guest file write test with count parameter

### PR DESCRIPTION
**Pull request by xiagao**
The check_guest_file test case only cover basic guest file write test,
need to do more test for it.
ID:1468524
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
@xiangchunfu Could you help review this patch?

**Comment by xiagao** [comment]
(1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_guest_file.q35: PASS (57.08 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_guest_file.q35: PASS (176.21 s)

**Comment by xiagao** [comment]
@xiangchunfu Could you help review this patch?

**Comment by xiangchunfu** [comment]
I reviewed this patch, It looks good for me. 
 (1/2) Host_RHEL.mX.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.X.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_guest_file.q35: PASS (172.27 s)
 (2/2) Host_RHEL.mX.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.X.0.x86_64.io-github-autotest-qemu.qemu_guest_agent.isa_serial.check_guest_file.q35: PASS (195.59 s)

Acked-by:Xiangchun Fu xfu@redhat.com

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -1043,7 +1047,42 @@ def _change_bl():
...
 
-        error_context.context("Close the opened file.", logging.info)
+        self.gagent.guest_file_close(ret_handle)
+
+        # create a new file by agent.
+        error_context.context("Create new file with mode 'w' and do file write"
+                              " test", logging.info)
+        ranstr = utils_misc.generate_random_string(5)
+        tmp_file = "%s%s" % (params["file_path"], ranstr)
+        ret_handle = int(self.gagent.guest_file_open(tmp_file, mode="w+"))
```
Comment (peixiu) [inline qemu/tests/qemu_guest_agent.py]: 1055 -1057 lines, and 1009-1011 lines, they are same, why need repeat 2 times here?

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: They are not the same one. Here create a new  clean file to do following test.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -931,7 +931,11 @@ def gagent_check_guest_file(self, test, params, env):
              read two bytes.
           c. seek from current position and offset is 2,read 5 bytes.
           d. seek from the file end and offset is -5,read 3 bytes.
-        5) close file
+        5) create another new file with mode "w" and write to it.
+        6) write test with count parameter.
+          a. write two counts bytes to guest file.
+          b. write ten counts bytes to guest file from the file end.
+          c. write mort than all counts bytes to guest file.
```
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: update,thanks

**Comment by peixiu** [comment]
Looks good to me~
Acked-by: Peixiu Hou phou@redhat.com

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -839,6 +818,27 @@ def gagent_check_reboot_shutdown(self, test, params, env):
...
+        # Use the last opened session to send cmd.
+        session = self._open_session_list[-1]
+
+        try:
+            session.cmd(self.params["gagent_fs_test_cmd"])
+        except aexpect.ShellTimeoutError:
+            logging.info("FS is frozen as expected,can't write in guest.")
+        else:
+            self.test.fail("FS is not frozen,still can write in guest.")
+
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: Why move this section without any changes?

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -1043,7 +1047,42 @@ def _change_bl():
...
+        self.gagent.guest_file_close(ret_handle)
+
+        # create a new file by agent.
+        error_context.context("Create new file with mode 'w' and do file write"
+                              " test", logging.info)
+        ranstr = utils_misc.generate_random_string(5)
+        tmp_file = "%s%s" % (params["file_path"], ranstr)
+        ret_handle = int(self.gagent.guest_file_open(tmp_file, mode="w+"))
+
+        error_context.context("Write two counts bytes to guest file.",
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: `two counts of byte`?

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -1043,7 +1047,42 @@ def _change_bl():
...
+        self.gagent.guest_file_seek(ret_handle, 0, 0)
+        _read_check(ret_handle, "he")
+
+        error_context.context("Write ten counts bytes to guest file.",
+                              logging.info)
+        self.gagent.guest_file_seek(ret_handle, 0, 2)
+        self.gagent.guest_file_write(ret_handle, content, 10)
+        self.gagent.guest_file_flush(ret_handle)
+        self.gagent.guest_file_seek(ret_handle, 0, 0)
+        _read_check(ret_handle, "hehello worl")
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: How about this?
```
content_check = ""
for cnt in [2, 10]:
    error_context.context("Write %s counts of byte to guest file." % cnt, logging.info)
    self.gagent.guest_file_seek(ret_handle, 0, len(content_check))
    self.gagent.guest_file_write(ret_handle, content, cnt)
    self.gagent.guest_file_flush(ret_handle)
    self.gagent.guest_file_seek(ret_handle, 0, 0)
    content_check += content[: int(cnt)]
    _read_check(ret_handle, content_check)
````

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: As  `def gagent_check_reboot_shutdown() function`(merged recently) separate fsfreeze/fsthaw function.

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: It should be two bytes. Will update.thanks

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: Good idea, thank you :)

**Comment by xiagao** [comment]
@vivianQizhu  updated. Thanks

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -1043,7 +1047,38 @@ def _change_bl():
...
+                              " test", logging.info)
+        ranstr = utils_misc.generate_random_string(5)
+        tmp_file = "%s%s" % (params["file_path"], ranstr)
+        ret_handle = int(self.gagent.guest_file_open(tmp_file, mode="w+"))
+
+        content_check = ""
+        for cnt in range(1, 10, 2):
+            error_context.context("Write %s bytes to guest file." % cnt,
+                                  logging.info)
+            self.gagent.guest_file_seek(ret_handle, 0, 2)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py]: I suppose this should not be a fixed 2 as your original design.

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: 2 means seek the position from the end.It has no relationship with count.

**Comment by xiagao** [comment]
(1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_guest_file.q35: PASS (62.80 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_guest_file.q35: PASS (214.48 s)

Also pass on py3

File: qemu/tests/qemu_guest_agent.py (hunk 6 above)
Comment (vivianQizhu) [reply qemu/tests/qemu_guest_agent.py]: @xiagao Ah, right, then it looks good to me now. Thanks.

**Pull request merged**